JPA Map Storage: Keycloak fails to start on CockroachDB

### Describe the bug

When any storage is configured to use the Map JPA provider, Keycloak with CockroachDB fails to start due to lack of support for computed columns in foreign keys, which are present in the authorization changesets:

```
Exception in thread "main" java.lang.RuntimeException: RESTEASY003325: Failed to construct public org.keycloak.services.resources.KeycloakApplication()
	at org.jboss.resteasy.core.ConstructorInjectorImpl.construct(ConstructorInjectorImpl.java:164)
	at org.jboss.resteasy.spi.ResteasyProviderFactory.createProviderInstance(ResteasyProviderFactory.java:2835)
...

Caused by: liquibase.exception.LiquibaseException: liquibase.exception.MigrationFailedException: Migration failed for change set META-INF/authz/permission/jpa-authz-permission-changelog-1.xml::authz-permission-1::keycloak:
     Reason: liquibase.exception.DatabaseException: ERROR: unimplemented: computed column "resourceserverid" cannot reference a foreign key
  Hint: You have attempted to use a feature that is not yet implemented.
See: https://go.crdb.dev/issue-v/46672/v21.2 [Failed SQL: (0) ALTER TABLE public.kc_authz_permission ADD CONSTRAINT authz_permission_fk_root_fkey FOREIGN KEY (resourceserverid) REFERENCES public.kc_authz_resource_server (id) ON DELETE CASCADE]

```

We might have to move all the foreign key constraints in the authz changelogs that reference a computed column to a changeset that runs only for PostgreSQL.

### Version

main

### Expected behavior

Keycloak should be able to process all new storage changelogs when CockroachDB is used.

### Actual behavior

Keycloak fails to start if any area is configured to use the new Map JPA store (which in Cockroach's case leads to all changelogs being processed)

### How to Reproduce?

_No response_

### Anything else?

_No response_

--- a/model/map-jpa/src/main/resources/META-INF/authz/permission/jpa-authz-permission-changelog-1.xml
+++ b/model/map-jpa/src/main/resources/META-INF/authz/permission/jpa-authz-permission-changelog-1.xml
@@ -58,6 +58,12 @@ limitations under the License.
             <ext:column jsonColumn="metadata" jsonProperty="fGrantedTimestamp"/>
         </ext:createJsonIndex>
         <addUniqueConstraint columnNames="realmid, resourceserverid, scopeid, resourceid, owner, requester"  tableName="kc_authz_permission"/>
+    </changeSet>
+
+    <!-- unimplemented on CockroachDB: computed column "XXX" cannot reference a foreign key
+         See: https://go.crdb.dev/issue-v/46672/v21.2 -->
+
+    <changeSet author="keycloak" id="authz-permission-2" dbms="postgresql">
         <addForeignKeyConstraint constraintName="authz_permission_fk_root_fkey"
                                  baseTableName="kc_authz_permission"
                                  baseColumnNames="resourceserverid"
@@ -65,6 +71,13 @@ limitations under the License.
                                  referencedColumnNames="id"
                                  onDelete="CASCADE"/>
 
+    </changeSet>
+
+    <changeSet author="keycloak" id="authz-permission-3" dbms="cockroachdb">
+        <!-- placeholder for foreign key on computed column to allow fast looking by resourceserverid -->
+        <createIndex tableName="kc_authz_permission" indexName="authz_permission_fk_root_fkey">
+            <column name="resourceserverid"/>
+        </createIndex>
     </changeSet>
 
 </databaseChangeLog>

--- a/model/map-jpa/src/main/resources/META-INF/authz/policy/jpa-authz-policy-changelog-1.xml
+++ b/model/map-jpa/src/main/resources/META-INF/authz/policy/jpa-authz-policy-changelog-1.xml
@@ -52,12 +52,6 @@ limitations under the License.
         <createIndex tableName="kc_authz_policy" indexName="authz_policy_type">
             <column name="type"/>
         </createIndex>
-        <addForeignKeyConstraint constraintName="authz_policy_fk_root_fkey"
-                                 baseTableName="kc_authz_policy"
-                                 baseColumnNames="resourceserverid"
-                                 referencedTableName="kc_authz_resource_server"
-                                 referencedColumnNames="id"
-                                 onDelete="CASCADE"/>
 
         <createTable tableName="kc_authz_policy_scope">
             <column name="policy_id" type="UUID">
@@ -118,8 +112,6 @@ limitations under the License.
             <column name="fk_root"/>
             <column name="name"/>
         </createIndex>
-        <!-- this is deferrable and initiallyDeferred as hibernate will first insert new entries and then delete the old by default -->
-        <addUniqueConstraint tableName="kc_authz_policy_config" columnNames="fk_root, name" deferrable="true" initiallyDeferred="true" />
     </changeSet>
 
     <changeSet author="keycloak" id="authz-policy-2" dbms="postgresql">
@@ -129,5 +121,25 @@ limitations under the License.
         <addUniqueConstraint tableName="kc_authz_policy_scope" columnNames="policy_id, scope_id" deferrable="true" initiallyDeferred="true" />
         <addUniqueConstraint tableName="kc_authz_policy_resource" columnNames="policy_id, resource_id" deferrable="true" initiallyDeferred="true" />
         <addUniqueConstraint tableName="kc_authz_policy_associated_policy" columnNames="policy_id, associated_policy_id" deferrable="true" initiallyDeferred="true" />
+
+        <!-- this is deferrable and initiallyDeferred as hibernate will first insert new entries and then delete the old by default -->
+        <addUniqueConstraint tableName="kc_authz_policy_config" columnNames="fk_root, name" deferrable="true" initiallyDeferred="true" />
+
+        <!-- unimplemented on CockroachDB: computed column "XXX" cannot reference a foreign key
+             See: https://go.crdb.dev/issue-v/46672/v21.2 -->
+        <addForeignKeyConstraint constraintName="authz_policy_fk_root_fkey"
+                                 baseTableName="kc_authz_policy"
+                                 baseColumnNames="resourceserverid"
+                                 referencedTableName="kc_authz_resource_server"
+                                 referencedColumnNames="id"
+                                 onDelete="CASCADE"/>
     </changeSet>
+
+    <changeSet author="keycloak" id="authz-policy-3" dbms="cockroachdb">
+        <!-- placeholder for foreign key on computed column to allow fast looking by resourceserverid -->
+        <createIndex tableName="kc_authz_policy" indexName="authz_policy_fk_root_fkey">
+            <column name="resourceserverid"/>
+        </createIndex>
+    </changeSet>
+
 </databaseChangeLog>

--- a/model/map-jpa/src/main/resources/META-INF/authz/resource/jpa-authz-resource-changelog-1.xml
+++ b/model/map-jpa/src/main/resources/META-INF/authz/resource/jpa-authz-resource-changelog-1.xml
@@ -56,12 +56,6 @@ limitations under the License.
             <column name="name"/>
         </createIndex>
         <addUniqueConstraint tableName="kc_authz_resource" columnNames="realmid, resourceserverid, name, owner"/>
-        <addForeignKeyConstraint constraintName="authz_resource_fk_root_fkey"
-                                 baseTableName="kc_authz_resource"
-                                 baseColumnNames="resourceserverid"
-                                 referencedTableName="kc_authz_resource_server"
-                                 referencedColumnNames="id"
-                                 onDelete="CASCADE"/>
 
         <createTable tableName="kc_authz_resource_scope">
             <column name="resource_id" type="UUID">
@@ -117,6 +111,22 @@ limitations under the License.
         <!-- see https://go.crdb.dev/issue-v/31632/v21.2 for the current status of the implementation -->
         <addUniqueConstraint tableName="kc_authz_resource_scope" columnNames="resource_id, scope_id" deferrable="true" initiallyDeferred="true" />
         <addUniqueConstraint tableName="kc_authz_resource_attribute" columnNames="fk_root, name, value" deferrable="true" initiallyDeferred="true" />
+
+        <!-- unimplemented on CockroachDB: computed column "XXX" cannot reference a foreign key
+             See: https://go.crdb.dev/issue-v/46672/v21.2 -->
+        <addForeignKeyConstraint constraintName="authz_resource_fk_root_fkey"
+                                 baseTableName="kc_authz_resource"
+                                 baseColumnNames="resourceserverid"
+                                 referencedTableName="kc_authz_resource_server"
+                                 referencedColumnNames="id"
+                                 onDelete="CASCADE"/>
     </changeSet>
-    
+
+    <changeSet author="keycloak" id="authz-resource-3" dbms="cockroachdb">
+        <!-- placeholder for foreign key on computed column to allow fast looking by resourceserverid -->
+        <createIndex tableName="kc_authz_resource" indexName="authz_resource_fk_root_fkey">
+            <column name="resourceserverid"/>
+        </createIndex>
+    </changeSet>
+
 </databaseChangeLog>

--- a/model/map-jpa/src/main/resources/META-INF/authz/scope/jpa-authz-scope-changelog-1.xml
+++ b/model/map-jpa/src/main/resources/META-INF/authz/scope/jpa-authz-scope-changelog-1.xml
@@ -48,6 +48,11 @@ limitations under the License.
             <column name="resourceserverid"/>
             <column name="name"/>
         </createIndex>
+    </changeSet>
+
+    <!-- unimplemented on CockroachDB: computed column "XXX" cannot reference a foreign key
+         See: https://go.crdb.dev/issue-v/46672/v21.2 -->
+    <changeSet author="keycloak" id="authz-scope-2" dbms="postgresql">
         <addForeignKeyConstraint constraintName="authz_scope_fk_root_fkey"
                                  baseTableName="kc_authz_scope"
                                  baseColumnNames="resourceserverid"
@@ -55,6 +60,13 @@ limitations under the License.
                                  referencedColumnNames="id"
                                  onDelete="CASCADE"/>
 
+    </changeSet>
+
+    <changeSet author="keycloak" id="authz-scope-3" dbms="cockroachdb">
+        <!-- placeholder for foreign key on computed column to allow fast looking by resourceserverid -->
+        <createIndex tableName="kc_authz_scope" indexName="authz_scope_fk_root_fkey">
+            <column name="resourceserverid"/>
+        </createIndex>
     </changeSet>
 
 </databaseChangeLog>

